### PR TITLE
[1037] Cleanup engines (torch.nn.modules and fowards functions)

### LINKED ISSUE #1037: Clean up of engines
### Is your feature request related to a problem? Please describe.

Currently, the engines (in `engines.py`) are not `torch.nn.modules` and not all of them implement a `forward` function. This complicates the code in `model.py` and leads to other unexpected issues. 

### Describe the solution you'd like

See above. Any apparent inconsistencies should also be fixed there.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### LINKED ISSUE #1106: Model refactoring breaks model loading
### Describe the task. It can be a feature, documentation, etc.

In issue #1037 the engines are refactored by implementing forward passes for every class. This breaks the checkpointing since:

- The state dictionary changes (since new modules are added)
- This gives issues when the 'old' weights are loaded for inference or training _continues_ with the new modules

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by yperugachidiaz**
## Description

Clean up of `engines.py` file by implementing forward pass for classes:
- EmbeddingEngine
- LocalAssimilationEngine
- Local2GlobalAssimilationEngine
- GlobalAssimilationEngine
- ForecastingEngine

Plus cleanup `embeddings.py`  by implementing forward pass for class:
- StreamEmbedTransformer

Fixed sharding when modules are called in `trainer.py`. 

Work in progress, issue should not be closed yet.
Upcoming: new issue #1106 to fix the now broken checkpoints (new modules change the structure).

## Issue Number

Closes #1037 
Closes #1106 

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by clessig** [comment]
@yperugachidiaz : For the review, it will be important to have some before-PR/after-PR loss curves. We had some issues before with gradient checkpointing where the code was formally correct (as far as we could tell) but still behaved differently under training.

File: src/weathergen/model/embeddings.py [hunk 1]
```diff
@@ -195,6 +193,13 @@ def forward_columns(self, x_in, centroids):
 
         return out.to(torch.float16)
 
+    def forward(self, x_in, centroids):
```
Comment (sophie-xhonneux) [inline src/weathergen/model/embeddings.py:195]: nice clean up

**Review by clessig — COMMENTED** [review]
Many thanks for the cleanup. Looks good but could we add a docstring to all forward functions. This will help us a lot going forward. I am happy to help with filling in some details but maybe you can start with it. Please also make sure to use our conventions (some of the docstring in engines.py use a different convention).

**Comment by clessig** [comment]
> The results can be found [here](https://gitlab.jsc.fz-juelich.de/hedgedoc/ZSVRYxvaTxm8fBLSwt7t6Q?both)

Many thanks. If it's just a few plots, you can it also post directly here.
